File: examples/44-section-management/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/editor"
)

func main() {
	if err := run(); err != nil {
		log.Fatalf("Smoke test failed: %v", err)
	}
	log.Println("Section Management Smoke Test: SUCCESS")
	log.Println(
		"To verify visually: Open examples/output/44_section_management.pptx in PowerPoint and switch to 'Slide Sorter' view.",
	)
}

func run() error {
	const outputDir = "examples/output"
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output dir: %w", err)
	}

	tmpDir, tempErr := os.MkdirTemp("", "gopptx-section-management-*")
	if tempErr != nil {
		return fmt.Errorf("create temp dir: %w", tempErr)
	}
	defer func() {
		if cleanupErr := os.RemoveAll(tmpDir); cleanupErr != nil {
			log.Printf("warning: failed to remove temp dir %s: %v", tmpDir, cleanupErr)
		}
	}()

	inputFile := filepath.Join(tmpDir, "section_test_input.pptx")
	outFile := filepath.Join(outputDir, "44_section_management.pptx")

	if err := createInputDeck(inputFile); err != nil {
		return err
	}

	edit, err := editor.OpenPresentationEditor(inputFile)
	if err != nil {
		return err
	}
	defer func() {
		if closeErr := edit.Close(); closeErr != nil {
			log.Printf("warning: failed to close editor: %v", closeErr)
		}
	}()

	if err := addAndRenameSections(edit); err != nil {
		return err
	}

	if err := edit.Save(outFile); err != nil {
		return fmt.Errorf("failed to save: %w", err)
	}

	if err := verifySections(outFile); err != nil {
		return err
	}
	return nil
}

func createInputDeck(inputFile string) error {
	deck := pptx.NewPresentationBuilder("Section Demo")
	deck.AddTitleSlide("Intro Slide")
	deck.AddTitleSlide("Detail 1")
	deck.AddTitleSlide("Detail 2")
	deck.AddTitleSlide("Appendix")
	if err := deck.WriteToFile(inputFile); err != nil {
		return err
	}
	return nil
}

func addAndRenameSections(edit *editor.PresentationEditor) error {
	log.Println("Adding 'Introduction' section...")
	if err := edit.AddSection("Introduction", []int{0}); err != nil {
		return fmt.Errorf("failed to add intro section: %w", err)
	}

	log.Println("Adding 'Core Content' section...")
	if err := edit.AddSection("Core Content", []int{1, 2}); err != nil {
		return fmt.Errorf("failed to add core section: %w", err)
	}

	log.Println("Adding 'Appendix' section...")
	if err := edit.AddSection("Appendix", []int{3}); err != nil {
		return fmt.Errorf("failed to add appendix section: %w", err)
	}

	log.Println("Renaming 'Appendix' to 'Back Matter'...")
	if err := edit.RenameSection("Appendix", "Back Matter"); err != nil {
		return fmt.Errorf("failed to rename section: %w", err)
	}
	return nil
}

func verifySections(outFile string) error {
	checkEdit, err := editor.OpenPresentationEditor(outFile)
	if err != nil {
		return fmt.Errorf("failed to open result: %w", err)
	}
	defer func() {
		if closeErr := checkEdit.Close(); closeErr != nil {
			log.Printf("warning: failed to close verification editor: %v", closeErr)
		}
	}()

	sections := checkEdit.Sections()
	log.Printf("Found %d sections:\n", len(sections))
	for _, s := range sections {
		log.Printf("- %s (ID: %s, Slides: %d)\n", s.Name, s.GUID, len(s.SlideIDs))
	}

	if len(sections) != 3 {
		return fmt.Errorf("expected 3 sections, got %d", len(sections))
	}
	if sections[2].Name != "Back Matter" {
		return fmt.Errorf("expected last section to be 'Back Matter', got %q", sections[2].Name)
	}
	return nil
}

File: examples/44-section-management/main.py
"""Demonstrate section management: adding and organizing slides into sections.

This example demonstrates:
- Creating a multi-slide presentation
- Adding named sections that group slide indices
- Retrieving sections and verifying structure
- Sections visible in PowerPoint Slide Sorter view
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation


def _build_slides(prs: Presentation) -> None:
    """Build the base slide content."""
    prs.add_title_slide("Intro Slide")
    prs.add_bullet_slide("Detail 1", ["First detail point", "Supporting information"])
    prs.add_bullet_slide("Detail 2", ["Second detail point", "More supporting info"])
    prs.add_bullet_slide("Appendix", ["Reference material", "Additional data"])


def main() -> None:
    """Create presentation demonstrating section management."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)
    output_path = output_dir / "44_section_management.pptx"

    with Presentation.new("Section Demo") as prs:
        _build_slides(prs)

        # Add sections grouping slides
        prs.add_section("Introduction", [0])
        prs.add_section("Main Content", [1, 2])
        prs.add_section("Appendix", [3])

        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    with Presentation(str(output_path)) as prs:
        sections = prs.sections
        print(f"Created {len(sections)} sections:")
        for section in sections:
            print(
                f"  - {section.get('Name', section.get('name', '?'))}"
                f" -> slide IDs {section.get('SlideIDs', section.get('slide_ids', []))}"
            )

    print("\n=== SUMMARY ===")
    print("3 sections: Introduction, Main Content, Appendix")
    print("Tip: Open in PowerPoint Slide Sorter view to see sections.")


if __name__ == "__main__":
    main()
